/*
 * Copyright 2015 Alexey Baranov <me@kotiki.cc>. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package squim;

service ImageOptimizer {
  rpc OptimizeImage(stream ImageRequestPart)
      returns (stream ImageResponsePart) {}
}

enum ImageType {
  JPEG = 0;
  PNG = 1;
  GIF = 2;
  WEBP = 3;
  UNKNOWN = 4;
}

message ImageData { bytes bytes = 1; }

message ImageRequestPart {
  enum WebPCompressionType {
    AUTO = 0;
    LOSSY = 1;
    LOSSLESS = 2;
    MIXED = 3;
  }

  message Meta {
    message WebPOptimizationParams {
      double quality = 1;
      int32 method = 2;
      WebPCompressionType compression_type = 3;
      bool record_stats = 4;
    }

    // Expected type of image. Just for bookeeping, actual image type will be
    // determined from image data.
    ImageType expected_type = 1;

    // Expected input size.
    uint64 content_length = 2;

    // Which format to optimize to. For now, only webp is supported.
    ImageType target_type = 3;

    bool try_strip_alpha = 5;
    bool strip_iccp = 6;
    bool strip_exif = 7;
    bool strip_xmp = 8;
    double min_photo_metric = 9;
    uint32 timeout_millis = 10;

    WebPOptimizationParams webp_params = 11;
  }

  oneof payload {
    Meta meta = 1;
    ImageData image_data = 2;
  }
}

message ImageResponsePart {
  enum Result {
    OK = 0;
    DECODE_ERROR = 1;
    ENCODE_ERROR = 2;
    REJECTED = 3;
    TIMEOUT = 4;
    CONTRACT_ERROR = 5;
  }

  enum ColorScheme {
    GRAY = 0;
    GRAY_ALPHA = 1;
    RGB = 2;
    RGBA = 3;
    YUV = 4;
    YUVA = 5;
    CMYK = 6;
  }

  message Meta {
    Result code = 1;
    string message = 2;
  }

  message Stats {
    ImageType original_image_type = 1;
    ColorScheme original_color_scheme = 2;
    uint32 iccp_size = 3;
    uint32 exif_size = 4;
    uint32 xmp_size = 5;
    bool iccp_stripped = 6;
    bool exif_stripped = 7;
    bool xmp_stripped = 8;
    double psnr = 9;
    uint32 num_frames = 10;
    ColorScheme output_color_scheme = 11;
    uint32 width = 12;
    uint32 height = 13;
    bool is_photo = 14;
    uint32 coded_size = 15;
  }

  oneof payload {
    Meta meta = 1;
    ImageData image_data = 2;
    Stats stats = 3;
  }
}
